src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+    {
+        parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_type_manager, $payment_type_manager, $payment_method_type_manager, $time);
+  
+        // Heartland SDK - Authentication
+        // https://developer.heartlandpaymentsystems.com/Documentation/authentication/#authentication
+        $config = new ServicesConfig();
+        $config->secretApiKey = $this->configuration['secret_key'];
+        $config->serviceUrl = "https://cert.api2.heartlandportico.com";
+        ServicesContainer::configure($config);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPublicKey()
+    {
+        // Used by PaymentMethodAddForm to send key to javascript
+        return $this->configuration['public_key'];
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function defaultConfiguration()
+    {
+        // Store Card Data will be implemented in iteration 2 multi-use tokenization.
+        return [
+            'public_key' => '',
+            'secret_key' => '',
+            // 'store_card_data' => 'FALSE',
+        ] + parent::defaultConfiguration();
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function buildConfigurationForm(array $form, FormStateInterface $form_state)
+    {
+        $form = parent::buildConfigurationForm($form, $form_state);
+  
+        // Matching schema in config/schema/commerce_heartland.schema.yml.
+        $form['public_key'] = [
+            '#type' => 'textfield',
+            '#title' => $this->t('Public Key'),
+            '#default_value' => $this->configuration['public_key'],
+            '#required' => true,
+        ];
+
+        $form['secret_key'] = [
+            '#type' => 'textfield',
+            '#title' => $this->t('Secret Key'),
+            '#description' => $this->t('Get your keys at <a href="https://developer.heartlandpaymentsystems.com/Account/KeysandCredentials" target="_blank">https://developer.heartlandpaymentsystems.com/Account/KeysandCredentials</a>.'),
+            '#default_value' => $this->configuration['secret_key'],
+            '#required' => true,
+        ];
+
+        // $form['store_card_data'] = [
+        //     '#type' => 'checkbox',
+        //     '#title' => $this->t('Store Card Data'),
+        //     '#description' => $this->t('Selet this option to store card data.  You must have multi-use token enabled on your Heartland Payment Systems account.'),
+        //     '#default_value' => FALSE,
+        // ];
+  
+        return $form;
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function validateConfigurationForm(array &$form, FormStateInterface $form_state)
+    {
+        parent::validateConfigurationForm($form, $form_state);
+
+        // Check if keys match the environment test/live
+        if (!$form_state->getErrors()) {
+            $values = $form_state->getValues($form['#parents'])['configuration']['commerce_heartland'];
+            $mode = $values['mode'];
```

Comment:
Is `mode` provided by Drupal Commerce?

Reply:
Yes, mode is part of the parent configuration.